### DISCUSSION TITLE
[Chapter 5] - Q & A

### DISCUSSION THREAD

**Discussion by ahngilwoong** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#57] [CHAPTER 5 스트림 활용 - Q&A]
[복습 퀴즈]

Map과 FlotMap에 대해 알아보기
map 연산은 각 요소를 1:1로 매핑한다 했고
flatMap 연산은 각 요소를 여러 개의 요소로 매핑한다 했다.

Q1. 
스트림을 통해 짝수 찾고 찾은 짝수 2로 나누는 과정인데 map과 flotMap중 뭘 써야하는가 이유는 무엇인가.

```java
List<Integer> numbers = Arrays.asList(1, 2, 3, 4, 5, 6);
List<Integer> examList1= numbers.stream()
                                    .filter(num -> num % 2 == 0)
                                    .map(num -> num / 2)
                                    .collect(Collectors.toList());

List<Integer> examList2 = numbers.stream()
                                    .filter(num -> num % 2 == 0)
                                    .flatMap(num -> Stream.of(num / 2))
                                    .collect(Collectors.toList());
```

**Comment by BvrPark** [comment]
Q1. 제 생각엔 둘다 사용가능하지 않을까 싶은데 둘 중 어떤게 더 낫냐(?)라고 물으시면 `map`인 것 같습니다.
왜냐하면 `map`은 각 요소에 대해서 1:1로 매핑을 하는데 위의 코드는 num / 2를 통해 값을 생성하는데 이는 기존의 numbers의 길이와 같아서?
`map`이 더 적절한 것 같습니다.

**Reply by Irisation23** [reply]
FlotMap 좀 킹받네요

**Comment by coalong** [comment]
> map 연산은 각 요소를 1:1로 매핑한다 했고
flatMap 연산은 각 요소를 여러 개의 요소로 매핑한다 했다.

이 표현은 혹시 책에 나와있는 건가요?! 
요소를 1:1 매핑과 요소를 여러 개의 요소로 매핑한다는 표현을 차원느낌으로 이해해도 될까요?!

A. 제 생각에는 둘 다 사용 가능하지만 위 코드에서 평면화가 필요없기 때문에 단순히 새로운 요소로 매핑하는 `map` 을 사용하는 게 더 효율적이지 않을까 ?? 생각합니다.!
